functions/access.js

```diff
@@ -0,0 +1,123 @@
+const admin = require("firebase-admin");
+const functions = require("firebase-functions");
+const { HttpsError } = require("firebase-functions/lib/providers/https");
+const { passwordsMatch } = require("./auth");
+const { uuid } = require("uuidv4");
+
+const checkIsValidToken = async (venueId, uid, token) => {
+  if (!uid) return false;
+
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const granted = await venue.ref.collection("accessgranted").doc(uid).get();
+  if (!granted.exists || !granted.tokens) {
+    return false;
+  }
+
+  if (Object.keys(granted.tokens).includes(token)) {
+    // Record that the token was checked
+    if (!granted.tokens[token].usedAt) {
+      granted.tokens[token].usedAt = [];
+    }
+    granted.tokens[token].usedAt.push(new Date().getTime());
+    await venue.ref.collection("accessgranted").doc(uid).set(granted);
+    return true;
+  }
+
+  return false;
+};
+
+const getAccessDoc = async (venueId, method) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const accessDoc = await venue.ref.collection("access").doc(method).get();
+  return accessDoc;
+};
+
+const isValidPassword = async (venueId, password) => {
+  const access = await getAccessDoc(venueId, "password");
+  if (!access.exists || !access.data().password) {
+    return false;
+  }
+  return passwordsMatch(access.data().password, password);
+};
+
+const isValidEmail = async (venueId, email) => {
+  const access = await getAccessDoc(venueId, "emails");
+  if (!access.exists || !access.data().emails) {
+    return false;
+  }
+  return access.data().emails.includes(email.trim().toLowerCase());
+};
+
+const isValidCode = async (venueId, code) => {
+  const access = getAccessDoc(venueId, "code");
+  if (!access.exists || !access.data().codes) {
+    return false;
+  }
+  return access.data().codes.includes(code.trim());
+};
+
+const createToken = async (venueId, uid, password, email, code) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  let granted = {};
+  let grantedDoc = await venue.ref.collection("accessgranted").doc(uid).get();
+  if (grantedDoc.exists) {
+    granted = grantedDoc.data();
+  }
+  if (!granted.tokens) {
+    granted.tokens = {};
+  }
+  // Record the first time the token was created
+  const token = uuid();
+  granted.tokens[token] = {
+    usedAt: [new Date().getTime()],
+    password: password ?? null,
+    email: email ?? null,
+    code: code ?? null,
+  };
```

Comment:
We probably shouldn't be modifying the object here. We can use object spread to replace keys later on if we want to 'update' parts of it (though I don't know that we even need that, see comment RE: update below)

Reply:
The object is not being modified here, it's being created. The keys are not being replaced. I'm not sure how spread operator will help us here.

Reply:
(sidenote: Rebasing and force pushing make code reviews a lot harder, as it breaks any code references that already existed)

The code above this is reading from the DB:
```
const granted = await venue.ref.collection("accessgranted").doc(uid).get();
```

Which is then being mutated in the lines I called out.

Where I said using spread would help would be something like the following, which avoids modifying the previous object:

```typescript
theFinalFuncCall({ ...granted, token: { etc }})
```

The "though not sure we even need that" is referring to just updating the explicit key we want to change, rather than overwriting the entire thing, as per https://github.com/sparkletown/sparkle/pull/1016#discussion_r559294516